test/e2e/upgrade_test.go

```diff
@@ -0,0 +1,413 @@
+      "umip",
+      "vaes",
+      "vme",
+      "vmx",
+      "vnmi",
+      "vpclmulqdq",
+      "vpid",
+      "waitpkg",
+      "x2apic",
+      "xgetbv1",
+      "xsave",
+      "xsavec",
+      "xsaveopt",
+      "xsaves",
+      "xtopology"
+    ],
+    "model": "12th Gen Intel(R) Core(TM) i9-12900H"
+  },
+  "firmware": {
+    "bios": {
+      "date": "04/01/2014",
+      "vendor": "SeaBIOS",
+      "version": "1.15.0-1"
+    }
+  },
+  "hostname": "localhost.localdomain",
+  "nics": [
+    {
+      "ip": "192.168.222.122",
+      "mac": "00:60:2f:31:81:01",
+      "model": "0x1af4 0x0001",
+      "name": "enp1s0",
+      "pxe": true
+    },
+    {
+      "ip": "fe80::570a:edf2:a3a7:4eb8%enp1s0",
+      "mac": "00:60:2f:31:81:01",
+      "model": "0x1af4 0x0001",
+      "name": "enp1s0",
+      "pxe": true
+    }
+  ],
+  "ramMebibytes": 4096,
+  "storage": [
+    {
+      "name": "/dev/disk/by-path/pci-0000:04:00.0",
+      "rotational": true,
+      "sizeBytes": 21474836480,
+      "type": "HDD",
+      "vendor": "0x1af4"
+    }
+  ],
+  "systemVendor": {
+    "manufacturer": "QEMU",
+    "productName": "Standard PC (Q35 + ICH9, 2009)"
+  }
+}
+`
+
+var _ = Describe("BMO Upgrade", func() {
+	var (
+		specName                  = "upgrade"
+		namespace                 *corev1.Namespace
+		bmoNamespace              string
+		bmoDeployName             string
+		bmcUser                   string
+		bmcPassword               string
+		bmcAddress                string
+		bootMacAddress            string
+		bmoKustomization          string
+		upgradeClusterProvider    bootstrap.ClusterProvider
+		upgradeClusterProxy       framework.ClusterProxy
+		bmh                       metal3api.BareMetalHost
+		kubeconfigPath            string
+		originalKubeConfigContext string
+		contextChangedBeforeTest  = false
+	)
+	BeforeEach(func() {
+		bmcUser = e2eConfig.GetVariable("BMC_USER")
+		bmcPassword = e2eConfig.GetVariable("BMC_PASSWORD")
+		bmcAddress = e2eConfig.GetVariable("BMC_ADDRESS")
+		bootMacAddress = e2eConfig.GetVariable("BOOT_MAC_ADDRESS")
+		bmoNamespace = "baremetal-operator-system"
+		bmoDeployName = "baremetal-operator-controller-manager"
+		bmoKustomization = e2eConfig.GetVariable("BMO_KUSTOMIZATION")
+		bmoUpgradeFromKustomization := e2eConfig.GetVariable("BMO_UPGRADE_FROM_KUSTOMIZATION")
+
+		if useExistingCluster {
+			kubeconfigPath = os.Getenv("KUBECONFIG")
+			if kubeconfigPath == "" {
+				kubeconfigPath = os.Getenv("HOME") + "/.kube/config"
+			}
+			var err error
+			originalKubeConfigContext, err = GetCurrentKubeconfigContextName(kubeconfigPath)
+			Expect(err).NotTo(HaveOccurred())
+			k8sUpgradeContextName := e2eConfig.GetVariable("KUBECONFIG_UPGRADE_CONTEXT_NAME")
+			if (k8sUpgradeContextName != originalKubeConfigContext) && (k8sUpgradeContextName != "") {
+				By(fmt.Sprintf("Using kubeconfig context: %s for upgrade tests", k8sUpgradeContextName))
+				err = SwitchKubeconfigContext(kubeconfigPath, k8sUpgradeContextName)
```

Comment:
We cannot switch contexts like this in the tests. It will break if we run tests in parallel (since all tests use the currently selected context). I'm sorry I guided you towards context, but I think we need to use separate kubeconfigs.